Avoid round-tripping URLs through URI instances

ulfjack opened the pull request:
Instead, keep the original URL object returned from the ClassLoader. This
retains any associated URLStreamHandler with the URL, which allows the
ClassLoader to return in-memory resources.

As an alternative to this change, projects that want to use in-memory
resources with Hazelcast could replace the URLStreamHandlerFactory, adding a
custom scheme for in-memory resources. However, that seems a more invasive
approach, with this change being strictly simpler.

Fixes #16846.

devOpsHazelcast commented:
[![CLA assistant check](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/pull/badge/not_signed)](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/hazelcast/hazelcast?pullRequest=16851) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/hazelcast/hazelcast?pullRequest=16851) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/check/hazelcast/hazelcast?pullRequest=16851) it.</sub>

devOpsHazelcast commented:
Can one of the admins verify this patch?

ulfjack commented:
- I just sent in the signed CLA.
- I tried running the checkstyle/spotbugs check, but that seems to fail due to pre-existing issues.

mmedenjak commented:
run-lab-run

mmedenjak commented:
Thank you for the contribution @ulfjack ! I'll try to find reviewers and we'll try to get it merged soon. In addition to that, we may think about backporting this to `4.0.z` branch (and possibly `3.12.z`) once it's approved and merged so it gets released with the next 4.0.x or 3.12.x patch release.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/52/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/52/Hazelcast-pr-builder-52.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal com.github.spotbugs:spotbugs-maven-plugin:3.1.12:check (default) on project hazelcast: failed with 2 bugs and 0 errors -> [Help 1]
--------------------------
</details>
</pre>

mmedenjak commented:
@ulfjack can you check spotbugs and checkstyle locally by running `mvn clean install -DskipTests -Pcheckstyle,spotbugs`?

Holmistr commented:
@ulfjack Actually, we have switched to a new CLA online signing process as the bot commented here: https://github.com/hazelcast/hazelcast/pull/16851#issuecomment-609778303 
Would you mind clicking the button there and signing it electronically? It won't bother you anymore :)

ulfjack commented:
Here's the output I get from ```mvn clean install -DskipTests -Pcheckstyle,spotbugs```:
```
[INFO] --- maven-checkstyle-plugin:3.1.1:checkstyle (default) @ hazelcast ---
[INFO] Starting audit...
[ERROR] /my/home/directory/hazelcast/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapPutAllMessageTask.java:50: Line is longer than 130 characters (found 153). [LineLength]
Audit done.
```

I haven't touched that file.

It looks like it gets further when I run the check at HEAD, but it doesn't pass either:
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Hazelcast Root 4.1-SNAPSHOT:
[INFO] 
[INFO] Hazelcast Root ..................................... SUCCESS [  2.379 s]
[INFO] hazelcast .......................................... FAILURE [01:46 min]
[INFO] hazelcast-spring ................................... SKIPPED
[INFO] hazelcast-build-utils .............................. SKIPPED
[INFO] hazelcast-all ...................................... SKIPPED
[INFO] modulepath-tests ................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:49 min
[INFO] Finished at: 2020-04-08T18:59:11+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.spotbugs:spotbugs-maven-plugin:3.1.12:check (default) on project hazelcast: failed with 4 bugs and 0 errors  -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :hazelcast
```

If you have a way to check the change I made specifically, I'm happy to fix issues with it. I'm not going to fix any other issues in your code base.

Let me also say that it seems problematic for the check to exit on the first violation rather than printing all of them - at least that way I could check if there are violations in the changes I made.

It looks like the checkstyle check passes for my change when I rebase to HEAD.

mmedenjak commented:
As you've noticed, you needed to rebase as the issue has already been fixed.

mmedenjak commented:
run-lab-run

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/60/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/60/Hazelcast-pr-builder-60.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal com.github.spotbugs:spotbugs-maven-plugin:3.1.12:check (default) on project hazelcast: failed with 2 bugs and 0 errors -> [Help 1]
--------------------------
</details>
</pre>

mmedenjak commented:
There are still spotbugs issues in this PR:
```
20:36:50 [ERROR] Invocation of java.net.URL.equals(Object), which blocks to do domain name resolution, in com.hazelcast.internal.util.ServiceLoader$URLDefinition.equals(Object) [com.hazelcast.internal.util.ServiceLoader$URLDefinition] At ServiceLoader.java:[line 252] DMI_BLOCKING_METHODS_ON_URL
20:36:50 [ERROR] Invocation of java.net.URL.hashCode(), which blocks to do domain name resolution, in com.hazelcast.internal.util.ServiceLoader$URLDefinition.hashCode() [com.hazelcast.internal.util.ServiceLoader$URLDefinition] At ServiceLoader.java:[line 260] DMI_BLOCKING_METHODS_ON_URL
```

Please take a look.

ulfjack commented:
Updated to avoid URL.equals and URL.hashCode. I also rebased on HEAD. However, I still get spotbugs reports outside of the code I touched:
```
[INFO] --- spotbugs-maven-plugin:3.1.12:check (default) @ hazelcast ---
[INFO] BugInstance size is 4
[INFO] Error size is 0
[INFO] Total bugs: 4
[ERROR] Nullcheck of parser at line 42 of value previously dereferenced in com.hazelcast.map.impl.JsonMetadataInitializer.createFromData(Data) [com.hazelcast.map.impl.JsonMetadataInitializer, com.hazelcast.map.impl.JsonMetadataInitializer] At JsonMetadataInitializer.java:[line 42]Redundant null check at JsonMetadataInitializer.java:[line 43] RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE
[ERROR] Nullcheck of parser at line 52 of value previously dereferenced in com.hazelcast.map.impl.JsonMetadataInitializer.createFromObject(Object) [com.hazelcast.map.impl.JsonMetadataInitializer, com.hazelcast.map.impl.JsonMetadataInitializer] At JsonMetadataInitializer.java:[line 52]Redundant null check at JsonMetadataInitializer.java:[line 53] RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE
[ERROR] Nullcheck of parser at line 111 of value previously dereferenced in com.hazelcast.query.impl.getters.AbstractJsonGetter.getValue(Object, String) [com.hazelcast.query.impl.getters.AbstractJsonGetter, com.hazelcast.query.impl.getters.AbstractJsonGetter] At AbstractJsonGetter.java:[line 111]Redundant null check at AbstractJsonGetter.java:[line 141] RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE
[ERROR] com.hazelcast.query.impl.predicates.CompositeIndexVisitor$Output doesn't override java.util.ArrayList.equals(Object) [com.hazelcast.query.impl.predicates.CompositeIndexVisitor$Output] At CompositeIndexVisitor.java:[line 1] EQ_DOESNT_OVERRIDE_EQUALS
```

mmedenjak commented:
run-lab-run

mmedenjak commented:
Thank you for your contribution @ulfjack ! This fix will be part of 4.1 release. If you wish to get it in 4.0.2 or even 3.12.7, I suggest backporting it to `4.0.z` and `3.12.z` branches. Happy hazelcasting!

cc @Holmistr
